src/rust/engine/grpc_util/src/retry.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::time::Duration;
+
+use futures::Future;
+use rand::{thread_rng, Rng};
+use tonic::{Code, Response, Status};
+
+fn is_retryable(status: &Status) -> bool {
+  matches!(
+    status.code(),
+    Code::Aborted
+      | Code::Cancelled
+      | Code::Internal
+      | Code::ResourceExhausted
+      | Code::Unavailable
+      | Code::Unknown
+  )
+}
+
+/// Retry a gRPC client operation using exponential back-off to delay between attempts.
+#[inline]
+pub async fn retry_call<T, C, F, Fut>(client: C, f: F) -> Result<Response<T>, Status>
+where
+  C: Clone,
+  F: Fn(C) -> Fut,
+  Fut: Future<Output = Result<Response<T>, Status>>,
+{
+  const INTERVAL_DURATION: Duration = Duration::from_millis(10);
+  const MAX_RETRIES: u32 = 3;
+  const MAX_BACKOFF_DURATION: Duration = Duration::from_secs(5);
+
+  let mut last_error: Option<Status> = None;
+
+  let mut num_retries = 0;
+  while num_retries < MAX_RETRIES {
+    // Delay before the next send attempt if this is a retry.
+    if num_retries > 0 {
+      let multiplier = thread_rng().gen_range(0..2_u32.pow(num_retries) + 1);
+      let sleep_time = INTERVAL_DURATION * multiplier;
+      let sleep_time = sleep_time.min(MAX_BACKOFF_DURATION);
+      tokio::time::sleep(sleep_time).await;
+    }
+
+    let client2 = client.clone();
+    let result_fut = f(client2);
+    match result_fut.await {
+      Ok(r) => return Ok(r),
+      Err(status) => {
+        if is_retryable(&status) {
+          last_error = Some(status);
+        } else {
+          return Err(status);
+        }
+      }
+    }
+
+    num_retries += 1
+  }
+
+  Err(last_error.unwrap())
```

Comment:
This should be safe as it will always be set by line 52, but would love a second opinion.

Reply:
Yes, I agree this is safe. However, could you please change it to be `.expect()` and say something like "remote cache retry logic did not set `last_error` properly! Please file an issue at https://github.com/pantsbuild/pants/issues"

I'm thinking about that current failure we have "panic while panicking". Even though using expect vs panic wouldn't address that iiuc, it's helpful to make our panics more informative.

Reply:
I agree that this is safe. If you really wanted to get rid of the `unwrap()` here, you could turn your `while` statement into a `loop` expression:
```rust
let error = loop {
    ...
    if num_retries == MAX_RETRIES || !is_retryable(&status) {
        break status;
    }
}

Err(error)
```

or even just use a `loop` which always returns from the function or continues to loop, rather than needing to store a `Option`. But I could certainly argue both sides of any of the approaches :)

Reply:
`--remote-cache-max-retries` (or `--remote-store-max-retries`) does sound sensible to me. Not sure we need it, but makes sense.

Reply:
I like that trick with the `loop` returning a value. Much cleaner.